File: backend/app/schemas/agent.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


AgentMode = Literal["workflow", "design"]
AgentRole = Literal["user", "assistant", "system"]
AgentActionKind = Literal["text_to_image", "image_to_image", "split_multi_view"]
AgentActionStatus = Literal["draft", "confirmed", "submitted", "failed", "cancelled"]


class AgentAssetRef(BaseModel):
    asset_id: str | None = None
    name: str | None = None
    storage_url: str | None = None
    preview_url: str | None = None


class AgentActionCard(BaseModel):
    id: str | None = None
    kind: AgentActionKind
    module_key: str
    title: str = Field(min_length=1)
    prompt: str | None = None
    params: dict[str, object] = Field(default_factory=dict)
    source_assets: list[AgentAssetRef] = Field(default_factory=list)
    source_image_urls: list[str] = Field(default_factory=list)
    editable_prompt: bool = False
    next_question: str | None = None


class AgentMemoryProposal(BaseModel):
    content: str = Field(min_length=1)
    memory_type: str = "preference"


class AgentConversationCreate(BaseModel):
    mode: AgentMode = "workflow"
    title: str | None = None


class AgentConversationResponse(BaseModel):
    id: str
    mode: AgentMode
    title: str
    current_stage: str | None = None
    status: str
    summary: str | None = None
    state: dict[str, object] | None = None
    created_at: datetime
    updated_at: datetime


class AgentMessageResponse(BaseModel):
    id: str
    conversation_id: str
    role: AgentRole
    content: str
    attachments: list[AgentAssetRef] = Field(default_factory=list)
    event: dict[str, object] | None = None
    created_at: datetime


class AgentActionResponse(BaseModel):
    id: str
    conversation_id: str
    kind: AgentActionKind
    module_key: str
    status: AgentActionStatus
    title: str
    prompt: str | None = None
    params: dict[str, object] = Field(default_factory=dict)
    source_assets: list[AgentAssetRef] = Field(default_factory=list)
    source_image_urls: list[str] = Field(default_factory=list)
    result_job_id: str | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class AgentConversationDetail(BaseModel):
    conversation: AgentConversationResponse
    messages: list[AgentMessageResponse]
    actions: list[AgentActionResponse]


class AgentMessageCreate(BaseModel):
    content: str = ""
    mode: AgentMode | None = None
    attachments: list[AgentAssetRef] = Field(default_factory=list)


class AgentDesignStateUpdate(BaseModel):
    design_brief: dict[str, object] | None = None
    selected_knowledge_cards: list[dict[str, object]] | None = None


class AgentActionConfirmRequest(BaseModel):
    prompt: str | None = None
    params: dict[str, object] | None = None
    source_assets: list[AgentAssetRef] | None = None
    source_image_urls: list[str] | None = None


class AgentActionConfirmResponse(BaseModel):
    action: AgentActionResponse
    job_id: str
    status: str
    message: str


class AgentGenerationResultRegister(BaseModel):
    action_id: str | None = None
    module_key: str
    image_url: str = Field(min_length=1)
    name: str | None = None


class AgentUserMemoryCreate(BaseModel):
    content: str = Field(min_length=1)
    memory_type: str = "preference"
    source_conversation_id: str | None = None


class AgentUserMemoryResponse(BaseModel):
    id: str
    memory_type: str
    content: str
    is_enabled: bool
    source_conversation_id: str | None = None
    created_at: datetime
    updated_at: datetime


class AgentUserMemoryUpdate(BaseModel):
    is_enabled: bool | None = None
    content: str | None = None

File: frontend/src/types/agent.ts
import type { GenerationJobStatus } from "./fusion";

export type AgentMode = "workflow" | "design";
export type AgentRole = "user" | "assistant" | "system";
export type AgentActionKind = "text_to_image" | "image_to_image" | "split_multi_view";
export type AgentActionStatus = "draft" | "confirmed" | "submitted" | "failed" | "cancelled";

export interface AgentAssetRef {
  asset_id?: string | null;
  name?: string | null;
  storage_url?: string | null;
  preview_url?: string | null;
}

export interface AgentConversation {
  id: string;
  mode: AgentMode;
  title: string;
  current_stage?: string | null;
  status: string;
  summary?: string | null;
  state?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface AgentMessage {
  id: string;
  conversation_id: string;
  role: AgentRole;
  content: string;
  attachments: AgentAssetRef[];
  event?: Record<string, unknown> | null;
  created_at: string;
}

export interface AgentAction {
  id: string;
  conversation_id: string;
  kind: AgentActionKind;
  module_key: string;
  status: AgentActionStatus;
  title: string;
  prompt?: string | null;
  params: Record<string, unknown>;
  source_assets: AgentAssetRef[];
  source_image_urls: string[];
  result_job_id?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface AgentConversationDetail {
  conversation: AgentConversation;
  messages: AgentMessage[];
  actions: AgentAction[];
}

export interface AgentMemoryProposal {
  content: string;
  memory_type: string;
}

export interface AgentUserMemory {
  id: string;
  memory_type: string;
  content: string;
  is_enabled: boolean;
  source_conversation_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface AgentActionConfirmResult {
  action: AgentAction;
  job_id: string;
  status: GenerationJobStatus;
  message: string;
}

export interface AgentGenerationResultRegister {
  action_id?: string | null;
  module_key: string;
  image_url: string;
  name?: string | null;
}

export interface AgentKnowledgeCard {
  id: string;
  category: string;
  title: string;
  content: string;
}

export interface AgentDesignOption {
  label: string;
  value: string;
  description?: string;
}

export interface AgentDesignState {
  design_brief: Record<string, unknown>;
  selected_knowledge_cards: AgentKnowledgeCard[];
  stone_analysis?: Record<string, unknown> | null;
  knowledge_cards: AgentKnowledgeCard[];
  latest_design_mode?: "text_to_image" | "gemstone_design";
  pending_design_options?: AgentDesignOption[];
  pending_design_question?: string | null;
  pending_design_option_source?: "llm" | "fallback" | "none" | string;
}
